File: Demoblaze_Testing_TestNG_On_BrowserStack/src/main/java/com/example/pages/ProductPage.java
package com.example.pages;

import com.example.config.Constants;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class ProductPage {
    WebDriver driver;
    WebDriverWait wait;

    @FindBy(xpath = Constants.ADD_TO_CART)
    WebElement addToCart;

    public ProductPage(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(20));
        PageFactory.initElements(driver, this);
    }

    public void addToCartAndAcceptAlert() {
        wait.until(ExpectedConditions.elementToBeClickable(addToCart));
        addToCart.click();
        wait.until(ExpectedConditions.alertIsPresent());
        Alert alert = driver.switchTo().alert();
        alert.accept();
    }
}

File: Demoblaze_Testing_Pytest_On_BrowserStack/pages/product_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from config import ADD_TO_CART


class ProductPage:
    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 20)

    def add_to_cart_and_accept_alert(self):
        """
        Clicks the 'Add to cart' button and handles the alert popup on Demoblaze.
        Waits explicitly for the alert to be present (important for BrowserStack mobile).
        """
        # Wait for Add to Cart button
        add_button = self.wait.until(EC.element_to_be_clickable((By.XPATH, ADD_TO_CART)))
        add_button.click()

        # Wait for alert (Demoblaze may delay this by a few seconds)
        try:
            WebDriverWait(self.driver, 10).until(EC.alert_is_present())
            alert = self.driver.switch_to.alert
            alert_text = alert.text
            alert.accept()
            return alert_text
        except TimeoutException:
            raise AssertionError("❌ Alert did not appear after clicking 'Add to cart'. "
                                 "Possible UI delay or test environment issue.")
